### PR TITLE
Short-lived token refresh support - part 2

### PR DESCRIPTION
**Pull request by xuyanke**
- Created `DBAccessTokenProvider` protocol to dynamically provide access token for API requests.
-- Default implementations are provided for long-lived token and short-lived token.
- Refactored API client classes to use `DBAccessTokenProvider` instead of static access token string.
- In `DBClientsManager`:
-- Updated API clients setup methods to use appropriate access token providers.
-- Added the `db_` prefix to private methods and group them all together.

### CODE REVIEW

File: Source/ObjectiveDropboxOfficial/Shared/Handwritten/DBClientsManager.m [hunk 1]
```diff
@@ -163,138 +105,201 @@ + (void)setupWithOAuthManager:(DBOAuthManager *)oAuthManager
...
-  [DBClientsManager removeAuthorizedClient:tokenUid];
-  [DBClientsManager removeAuthorizedTeamClient:tokenUid];
++ (void)db_removeAllAuthorizedClients {
+  @synchronized(self) {
+    [s_tokenUidToAuthorizedClients removeAllObjects];
+  }
+}
+
++ (void)db_removeAllAuthorizedTeamClients {
+  @synchronized(self) {
```
[the remark this answers is not in the document]
Comment (xuyanke) [reply Source/ObjectiveDropboxOfficial/Shared/Handwritten/DBClientsManager.m:238]: I have not considered to replace this usage in this PR though, but sounds like a potential area to improve for the future.

File: Source/ObjectiveDropboxOfficial/Shared/Handwritten/DBClientsManager.m [hunk 2]
```diff
@@ -163,138 +105,201 @@ + (void)setupWithOAuthManager:(DBOAuthManager *)oAuthManager
...
+    [self db_addAuthorizedTeamClientWithToken:accessTokens[i] setAsDefault:i == 0];
+  }
+}
+
++ (void)db_resetClients {
+  [DBClientsManager db_setAuthorizedClient:nil];
+  [DBClientsManager db_setAuthorizedTeamClient:nil];
+
+  [DBClientsManager db_removeAllAuthorizedClients];
+  [DBClientsManager db_removeAllAuthorizedTeamClients];
```
[the remark this answers is not in the document]
Comment (xuyanke) [reply Source/ObjectiveDropboxOfficial/Shared/Handwritten/DBClientsManager.m:271]: Agreed, thanks for pointing out!

File: Source/ObjectiveDropboxOfficial/Shared/Handwritten/OAuth/DBAccessTokenProviderImpl.m [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+    // Ensure subsequent calls don't initiate more refresh requests, if one is in progress.
+    BOOL refreshInProgress = [self db_refreshInProgress];
+    [self->_completionBlocks addObject:completion];
+    if (!refreshInProgress) {
+      __weak typeof(self) weakSelf = self;
+      [self->_tokenRefresher refreshAccessToken:self->_token
+                                   scopes:@[]
+                                    queue:self->_queue
+                               completion:^(DBOAuthResult *result) {
+        [weakSelf db_handleRefreshResult:result];
```
[the remark this answers is not in the document]
Comment (rafe-g) [reply Source/ObjectiveDropboxOfficial/Shared/Handwritten/OAuth/DBAccessTokenProviderImpl.m]: Right, it dispatches back on the queue passed in. https://github.com/dropbox/dropbox-sdk-obj-c/pull/264/files#diff-7a40580386a62f3cc30c123bb74b44cdR101

I think the problem here is not that. Its that the fact that the dispatch is not happening on the barrier. _token.accessToken is assessed on a concurrent queue. It must be safely written by dispatching on barrier.

File: Source/ObjectiveDropboxOfficial/Shared/Handwritten/OAuth/DBAccessTokenProviderImpl.m [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+
+    // Ensure subsequent calls don't initiate more refresh requests, if one is in progress.
+    BOOL refreshInProgress = [self db_refreshInProgress];
+    [self->_completionBlocks addObject:completion];
+    if (!refreshInProgress) {
+      __weak typeof(self) weakSelf = self;
+      DBOAuthCompletion refreshCompletionBlock = ^(DBOAuthResult *result) {
+        typeof(self) strongSelf = weakSelf;
+        if (strongSelf) {
+          dispatch_async(strongSelf->_queue, ^{
```
Comment (rafe-g) [inline Source/ObjectiveDropboxOfficial/Shared/Handwritten/OAuth/DBAccessTokenProviderImpl.m:76]: I think this has to be on the barrier. See outdated convo. 

> Right, it dispatches back on the queue passed in. 

> https://github.com/dropbox/dropbox-sdk-obj-c/pull/264/files#diff-7a40580386a62f3cc30c123bb74b44cdR101

> I think the problem here is not that. Its that the fact that the dispatch is not happening on the barrier. 

> _token.accessToken is assessed on a concurrent queue. It must be safely written by dispatching on barrier.

Comment (xuyanke) [reply Source/ObjectiveDropboxOfficial/Shared/Handwritten/OAuth/DBAccessTokenProviderImpl.m:76]: oops, you are correct, all writes need to happen with barrier. thanks for catching this.

File: Source/ObjectiveDropboxOfficial/Shared/Handwritten/OAuth/DBAccessTokenProviderImpl.m [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+  BOOL dateHasPassed = fiveMinutesBeforeExpire.timeIntervalSinceNow < 0;
+  return dateHasPassed;
+}
+
+- (BOOL)db_refreshInProgress {
+  return _completionBlocks.count > 0;
+}
+
+- (void)db_handleRefreshResult:(DBOAuthResult *)result {
+  dispatch_barrier_async(_queue, ^{
```
Comment (xuyanke) [inline Source/ObjectiveDropboxOfficial/Shared/Handwritten/OAuth/DBAccessTokenProviderImpl.m:100]: @rafe-g, I updated this method to happen on barrier.

**Pull request merged**